server/etcdserver/api/v3rpc/grpc.go

```diff
@@ -91,7 +89,43 @@ func Server(s *etcdserver.EtcdServer, tls *tls.Config, interceptor grpc.UnarySer
 	pb.RegisterMaintenanceServer(grpcServer, NewMaintenanceServer(s, healthNotifier))
 
 	// set zero values for metrics registered for this grpc server
-	serverMetrics.InitializeMetrics(grpcServer)
+	metrics.InitializeMetrics(grpcServer)
 
 	return grpcServer
 }
+
+func getServerMetrics(metricType string, logger *zap.Logger) *grpc_prometheus.ServerMetrics {
+	// ensures only basic or extensive for testing
+	if metricType != "extensive" {
+		metricType = "basic"
+	} else {
+		metricType = "extensive"
+	}
+	metricsCacheMu.RLock()
```

Comment:
I think we don't need read mode here. just lock it, create new one, register it and then unlock.